### ISSUE TITLE
Make it easier to control the sequencing of Tasks

### ISSUE THREAD

**Issue by mighdoll** [sbt/sbt#1001]
As Eugene [also discussed](https://groups.google.com/forum/#!topic/simple-build-tool/1c7unObqjJ8), it's awkward to control the sequencing of tasks in sbt builds.  Let's see if we can find a simple control structure.

One common use case for task sequencing is for integration testing:
  Task 1: Start a test server / wait for it to come on line
  Task 2: run integration tests
  Task 3: stop servers

Controlling remote servers is another use case for task sequencing, and described in #1000.

I wonder if we could imitate the Future apis for Tasks (e.g. Future.sequence), and possibly use the scala async macro..

**Comment by jsuereth** [comment]
Funny enough, we have this in 2 chapters of Sbt In Action.

The general pattern we use (if, e.g. you have a single task which embodies all the functionality you need to wrap) is this:

``` scala

trait ServerStarter {
  def start(): Unit
  def stop(): Unit
}


val testServerStarter = taskKey[ServerStarter]("Creates an object which can start and stop servers.")

testServerStarter := ....

test in IntegrationTest := {
  for {
     service <- testServerStarter.taskValue
     _ = service.start()
    results <- (executeTests in IntegrationTest).taskVaue.andFinally(service.stop())
  } yield Test.reportResults(results)
}
```

Caveat that:
1. I'm using sbt 0.13.2 syntax (taskValue) rather than the old sbt 0.12 APIs like we do i thebook.
2. I may have mispellings, etc., but I hope the results are plain.
3. ALL The behavior you want wrapped in start/stop of test server must be in a SINGLE task and not strewn through multiple taskkeys, or you can't directly hook the Task API like this.   You'd have to ensure the flow of task execution between all the individual tasks _always_ occurs so it can make it to your "close" operation.  You also have to wire start into the first task, and end into the last task, and then you're not guaranteed any safety.

Basically, right now for-expressions are the best way to do sequencing.  I agree that it'd be nice to try to make things better. 

And, in actuality, the async macro is similar to the current task macro.  Things are parallel by default.    The macro makes up for Scala's like of "parallel" (applicative) syntactic sugar, while for-expressons represent the sequencing sugar (moandic).

So, while I'd love to have a better answer for you, I think the above pattern may be what you need to ape for the near term.

**Comment by harrah** [comment]
To clarify a potential misunderstanding, async is not sequential and is not parallel by default.  It unfortunately can't be used here because it has a different model than sbt tasks.

**Comment by mighdoll** [comment]
Josh, I just downloaded the current copy of sbt in action, feel free to point me to the right section.  I'm not yet able to follow your example above.

I like the trick of decomposing test into executeTests and Test.reportResults.  I think that answers one of my pending SO sbt questions...

If I could get your example to compile as you intend, would the macro that handles .value [1] arrange to make starter.value complete before executeTests is executed?  I'd have thought that that 'test' would have two parallel dependencies, 'starter' and 'executeTests'.

In any case, I've not been able to string two tasks together in a for expression.  tips?  e.g. this doesn't compile:

```
lazy val testTask2: Initialize[Task[Unit]] = Def.task {
  for {
    results <- (executeTests in IntegrationTest)
    results2 <- (executeTests in IntegrationTest)   // this doesn't compile
  } yield {
    Tests.showResults(streams.value.log, results, "missing tests?")
  }
}
```

[1] I presume you mean .value, not .taskValue.  Or is .taskValue something new?

**Comment by mighdoll** [comment]
Here's a stab at sequencing tasks via dependsOn.  It seems to work...

```
  // toy example of task sequencing

lazy val task1 = Def.task { println("one") }
lazy val task2 = Def.task { println("two") }
lazy val task3 = Def.task { println("three") }

lazy val testTask4: Initialize[Task[Unit]] = Def.taskDyn {
  SbtUtil.sequence(task1, task2, (test in IntegrationTest).toTask, task3)
}


/** Return a task that will execute the provided tasks sequentially */
def sequence(tasks: Initialize[Task[Unit]]*): Initialize[Task[Unit]] = {
  tasks.reverse.reduceLeft { (a, b) =>
    val combinedTask: Initialize[Task[Unit]] = {
      a.zip(b).apply { case (task1, task2) =>
        task1.dependsOn(task2)
      }
    }

    combinedTask
  }
}
```

If the above direction looks promising, it'd need some help generalizing the type parameters.  And I'm not sure what happens with error handling.  Hmm.. it'd be nice to have an andFinally/andThen for tasks too.

**Comment by harrah** [comment]
Implementing sequence is straightforward, you just need `flatMap` (or `taskDyn`) just like witih futures:

``` scala
def sequence(tasks: List[Initialize[Task[Unit]]]): Initialize[Task[Unit]] =
  tasks match {
    case Nil => Def.task{ () }
    case x :: xs => Def.taskDyn { val _ = x.value; sequence(xs) }
  }
```

However, this doesn't prevent something else from depending on `task3` and it executing sooner than `task1`, for example.  So, the task engine should also check that this sequence is respected.

**Comment by mighdoll** [comment]
> Implementing sequence is straightforward, you just need flatMap (or
> taskDyn) just like witih futures:
> 
> def sequence(tasks: List[Initialize[Task[Unit]]]): Initialize[Task[Unit]] =
>   tasks match {
>     case Nil => Def.task{ () }
>     case x :: xs => Def.taskDyn { val _ = x.value; sequence(xs) }
>   }
> 
> Pretty!
> 
> However, this doesn't prevent something else from depending on task3 and
> it executing sooner than task1, for example. So, the task engine should
> also check that this sequence is respected.
> 
> As an interim thing, I suppose the dependsOn version would inform the task
> engine of the appropriate dependency.

An initial question is whether sequence() should be part of the desired
api.  I suppose the existence of Future.sequence argues for it..

**Comment by harrah** [comment]
Hmm, what do you mean by "inform the task engine of the appropriate dependency".

I think `sequence` is a good and sufficient API for the use cases I've seen so far.  It doesn't provide an imperative API, but I don't know that's bad.  I don't think I'd want to have two similar looking APIs with such different behaviors.

**Comment by mighdoll** [comment]
> Hmm, what do you mean by "inform the task engine of the appropriate dependency".

I was probably misinterpreting your comment: 

> So, the task engine should also check that this sequence is respected. 

I don't know how the task engine works (seems nifty though, is there discussion of it anywhere?).  I was imagining that a) dependsOn dependencies were visible to the task engine and taskDyn ones were not, and b) the engine might recognize a cycle if it had more complete knowledge of the task dependency graph (imagine the sequence is 1,2,3 but there's also a dependency from 1 on 3).

> I think sequence is a good and sufficient API for the use cases I've seen so far

I think sequence would be handy.  Presumably sequence needs to work for Task[_], not just Task[Unit], but that didn't compile when I briefly tried it.  

Is there a way to use AndThen or AndFinally for tasks?  That would also seem handy for cases like the stopServer one above. 

I take it it's not feasible to use map / flatmap / filter and for comprehensions for sequencing as one might for Futures.  Out of interest, why is that?

>  I don't think I'd want to have two similar looking APIs with such different behaviors.

It's nifty that the task.value approach is parallel by default.  I'm sure some people will be bitten, though.  The expectation that imperative looking code is serially executed will be hard to shake.  Reminders in the documentation, release notes, etc. will help innoculate against those expectations.

**Comment by eed3si9n** [comment]
I'm skeptical of both `for` comprehension and `sequence(...)`. They are monadic zombie virus talking. For those who are infected, it may be ok, but if that's the case why bother with any DSL? The entire build definition can be a giant `for` comprehension running State-sponsored Reader/Writer/whatever.

Like I wrote in [Concurrency semantics (aka task sequencing) in 0.13 syntax](https://groups.google.com/d/msg/simple-build-tool/1c7unObqjJ8/3Wr30wMWKwEJ), 0.13 syntax "_almost_ feels like calling plain functions." That's the beauty and the beast of it. Someone who are used to sequential semantics would be like "woo startServer.value is so much easier to understand!" and then it's "gotcha! it's not sequential."

### async

Compare and contrast with SIP-22 async/await vs plain `Future[T]` that @jsuereth brought up. My `integrationTest` example in async would look like:

``` scala
def integrationTest: Future[Unit] = async {
  await {startServer}
  await {test}
  await {stopServer}
}
```

If you called plain startServer it's in asynchronous context, but you wrap it with `await {...}` the macro flatMaps it for you, which semantically is same as blocking till the call comes back.

### sequential

So one idea I mentioned was `ordered`. Let's call it `sequential` instead:

``` scala
integrationTest := sequential {
  // setup
  startServer.value
  // middle
  val jn = (jarName in assembly).value // depends on hello and version
  val h = hello.value
  val v = version.value
  doSomething(h, jn, v)
  // teardown
  stopServer.value
}
```

The difference between plain `Future[T]` and a setting is that it tracks setting dependencies. This introduces @harrah's comment ("However, this doesn't prevent something else from depending on `task3` and it executing sooner than `task1`"). IMHO that's perfectly ok. In the above code `(jarName in assembly)` depends on `hello` and `version`. You just get the cached value when you evaluate `hello.value` the second time.

The point of `sequential` is that it blocks at each tasks, so given no inter-dependencies, the build author can control the timing of the task firing. In this case, `startServer.value` fires and it blocks. If such macro existed, there would be much rejoicing.

### transaction

For more strict ordering I proposed `transaction` block. People say Monads are semicolons, but by twist of fate, we do have semicolons in sbt that cannot be encoded. I'm talking about batch mode:

```
$ sbt startServer; task1; task2; task3; stopServer
```

This is perfectly acceptable use of sbt, yet we can't write it as a task. We also can't emulate human typing those five tasks in a row from shell. Encoding that was the motivation behind transaction idea:

``` scala
integrationTest := transaction {
  startServer.value
} andThen { _ =>
  task1.value
} andThen { _ =>
  task2.value
} andThen { _ =>
  task3.value
} andThen { _ =>
  stopServer.value
}
```

or maybe we can let the user write it as:

``` scala
integrationTest := transaction {
  transaction {startServer.value}
  transaction {task1.value}
  transaction {task2.value}
  transaction {task3.value}
  transaction {stopServer.value}
}
```

If it's good enough for humans to type it in, it should be good enough to script it.

**Comment by jsuereth** [comment]
@eed3si9n You definitely have a point there.  I actually like the idea of a 'sequential' tag instead.  Again, this biggest "gotcha" in this kind of programming is if you assign one of the subsidiary tasks (to the sequential flow) to a Key and run any of those _in isolation_ or _use them in other tasks_.  

A great example of this is actually what @mighdoll has above.  He's using the `test in IntegrationTest` task which happens _after_ the actual execution, which is done in `executeTests in IntegrationTest`.   If you grab the wrong "Task[_]" object in your sequential flow, you almost need to have more exposure to the underlying model to know what will happen.

@mighdoll What you have written there may appear to work, but isn't 100%.  The safest thing to do is grab `executeTests` which actually performs the work, rather than the `test` which is pretty much just `Tests.printResults(executeTests.value)`.

**Comment by jsuereth** [comment]
@eed3si9n /all ALso for reference, Gradle's full task API is somewhat verbose:

``` groovy
task taskY {
    finalizedBy "taskX"
}
```

as an example.   In any case, I kind of like denoting "sequential" behavior explicitly, and maybe doing failure handling off the side, like:

``` scala
key := {
   val args = key.value
   val otherStuff = key.value
    sequentially {
       val server  = startServer.value
       try Tests.printResults((executeTests in IntegrationTest).value)
       finally {  stop(server) }
   }
}
```

i.e. The 'sequentially' block combined with normal try-catch can lift into an "andFinally" call on the task.   

Granted, I'm pie-in-the-skying here.  The practical reality of dealing with macros and AST transformations says the more limited we keep our rewriting, the easier it'll be to maintain.  Right now that means we're in Monad-land until we have a chance to spend the large time investment required for stable macros.

In any case, please keep the ideas flowing. Perhaps we should restart the sbt-dev thread?

**Comment by harrah** [comment]
I understand the sbt macro it looks like imperative code.  I'm ok with needing to educate users on that.  The sbt model is to have your side effects be local to your task so that as long as dependencies are satisfied, the task can be executed whenever.  The win is parallel by default and enabling faster builds in practice.

`scala/async` can't be reused because it targets the `Promise/Future style`, which doesn't preserve enough information or have the right model.  I'm not personally reimplementing `async` and then maintaining another large macro system either.  These things are huge time sinks.  So, I don't see it ever being implemented even if it were a somewhat reasonable idea.

A macro in the `scala/async` style has limitations as well.  @jsuereth's example doesn't currently work in async because `await` within `try/catch/finally` isn't supported.  (I don't know if it is fundamental- it might just not be implemented yet, but this style of macro is not at all trivial.)  Yes, if you do something so basic like your example, there is no problem, but then there isn't any advantage over sequence.  I don't think users have difficulty with what this does:

``` scala
executeInOrder(a, b, c)
```

So we call it `executeInOrder`.  There's nothing complicated there.  What is problematic is not if `jarName` depends on `name` and `version`, but if it depends on `stopServer`.  That is what I want to extend the task engine to verify isn't the case.

As for transactions, there are plenty of things you cannot write as a task.  That's what commands are for.  A task has the invariant that it is executed at most once within a task execution.  Semicolon is outside the task system because it breaks that invariant.

**Comment by mighdoll** [comment]
I'm still catching up with all three of you, but my reactions FWIW:

My mental model is that build developers are part time, very intermittent and sometimes reluctant users of any build system's language and APIs.  The less new they have to learn, the better.  The task macro is a step forward exactly so that they don't have to learn as much flatMap/apply/operator trickery.  Reusing what developers know from other contexts (scala in general, and Future/for comprehensions perhaps in this case) is a big win where that's possible.

sequence/executeInOrder works for my the use cases so far.  A warning if other dependencies break the sequence would be very handy.  Regarding naming, executeInOrder sounds more clear to me for this case, though consider sequence if it's feasible as a direction to re-use more of the Future api.

The 'sequential/sequentially' macro sounds nice for the cases I have for sequencing tasks, nicer than than sequence().  I don't have enough experience with the new system though to argue for the weight of a new macro.  Also I've seen a fair bit of weirdness in the eclipse IDE around macros, until that evolves I'm not eager for more macros.  (Type hovers are wrong, navigate to declaration doesn't work.  Known issues?)

What's the recommended andThen/andFinally solution for tasks?

> A great example of this is actually what @mighdoll has above. He's using the test in IntegrationTest task which happens after the actual execution, which is done in executeTests in IntegrationTest. If you grab the wrong "Task[_]" object in your sequential flow, you almost need to have more exposure to the underlying model to know what will happen.
> What you have written there may appear to work, but isn't 100%.

Wow, that sounds a bit scary...  I'm not sure which code you're referring to, but I'm guessing you mean this code:

```
/** run integration tests and then stop the test server */
lazy val testAndStop = Def.taskDyn {
  val _ = (test in IntegrationTest).value
  stop in testService
}
```

Can you be specific about what isn't 100% and why?  I think you're suggesting that a task (like the test task) might schedule some tasks (like executeTests) as a side effect somehow and that those side effects might be scheduled later.  Or maybe you mean that the returned task from taskDyn isn't guaranteed to execute after the body?

**Comment by jsuereth** [comment]
@mighdoll So, the testAndStop task is "fine" in the sense that if I call "it:test" it will run my tests, then stop, sort of.  If the tests result is an error, then (test in IntegrationTest).value _won't_ continue, because the result is an error, so my service is never stopped.

Also, if I just call `it:executeTests` then, again, I'll execute tests but never stop the service.  The ideal workaround is to have _one_ piece of the task-puzzle encapsulate the logic:
1. Start the server
2. Run the tests
3. Stop the server.

In this case `executeTests` is the `Task[_]` which will execute tests and return the results (be they success or failure).   What you want to do is take this `Task[_]` which executes all tests, and tie in your 'start/stop' server into it.

So in this case, it's easier to do something like:

``` scala
val startServerTask = Def.taskDyn { ... }

def stopServer(server: ???): Unit = ....

executeTests in IntegrationTest := {
  val originalTask = (executeTets in IntegrationTest).taskValue

  for {
     server <- startServerTask 
     result <- originalTask.andFinally(stopServer(server))
  } yield result
}
```

What this does is take the _original_ defintiion of the `executeTests` task and "wraps" it in an context such that
1. The startServerTask is called.  If this fails, the whole task fails.
2. The originalTask is executed and _always_ calls the stopServer method.
3. The result (Test results) are still returned.

As an alternative, you could imagine doing:

``` scala
val startServerTask = Def.taskDyn { ... }

val stopServerTask = Def.taskDyn { ... }

test in IntegrationTest := {
  executeInOrder(startServerTask, (test in IntegrationTest).taskValue, stopServerTask)
}
```

However, in this scenario, if something in startServerTask _accidentally_ depends on `executeTests in IntegrationTest` then the tests will have already executed _before_ you even start the server.  

The only guaranteed "ordered" behavior we have in this scenario is the behavior stored in `startServerTask`, `stopServerTask` and `test in IntegrationTest` that isn't _also_ defined in some dependent key.   For testing, the only behavior in defined in `test in IntegrationTest` is   `Tests.reportResult(executeTests.value)`.   This means if you accdientally depend on `executeTests` (indirectly or directly) then the tests will already have executed and the only enforced sequence we have is that the server is started _before_ we report all the failures.

That said, most likely, you do not have this dependency in practice, so things work as you expect.

This is my hesitance in evaluating @eed3si9n's improvement suggestions.  The general idea is awesome and can drastically improve everyone's life.  As you state, most build devs are not full time and need to be able to quickly get in and get out.

I wonder how much of a corner we're painted in when Task behavior is fragmented across keys though.  The async-by-default is by design to try to promote the best possible usage experience, since builds are used far more often than they are "designed".  

In any case, hope that cements some of the issues we're thinking through.   It could be in practice it's not a concern.  However, when working on Sbt in Action Chapter 6, we ran into this issue directly.   I'm still working on Chapter 9 where we go into more depth on the solutions and what's happening, but the use case of "run this server during integration tests" is very common.  It's also common people mistake a "user-level" task for the task which actually performs the work.  `test` being an example here, or `package`.   The actual "package a jar of my code" task is `packageBin in Compile` IIRC.  However, `package` is a convenience for users, and still useful for people plugging into a build, but not when you want to enforce ordering.

**Comment by mighdoll** [comment]
> If the tests result is an error, then (test in IntegrationTest).value won't continue, because the result is an error, so my service is never stopped.

Error handling stops the whole task, and so later tasks don't run.  OK, I get that and I see why that means the server might not be stopped if there's an error.

Would andThen/andFinally for tasks be the right fix for this?  (for a workaround I realize we need use what we have)

> The ideal workaround is to have one piece of the task-puzzle encapsulate the logic:
> What you want to do is take this Task[_] which executes all tests, and tie in your 'start/stop' server into it.
> I wonder how much of a corner we're painted in when Task behavior is fragmented across keys though.

Tasks are the 'atoms' that sbt build customizers are effectively encouraged to use for everything (because they have access to settings).  I suppose the core grammar of tasks is: run some in parallel, run some sequentially, attach task output to input, try/catch, attach task to key.  I'd like to imagine we get to a complete grammar around tasks.  Imagining I'm going to be explaining the task model, does that cover the core?  Also, the current andFinally seems (so far to me) to require dropping to a different level of abstraction since it works on function not task.  

> Chapter 9 where we go into more depth on the solutions and what's happening,

Looking forward to it.  I read all your books :-).

> It's also common people mistake a "user-level" task for the task which actually performs the work.
> is a convenience for users, and still useful for people plugging into a build, but not when you want to enforce ordering.

A category of user-level tasks that you should know but not use in build customizations sounds tricky to explain to people.  I hope we don't need to go there.  Customerizers may want to go deeper, of course, to e.g. inspect the success of executeTests.  But I hope there aren't tasks that would be unwise to use in a customization.

```
executeTests in IntegrationTest := {
  val originalTask = (executeTets in IntegrationTest).taskValue

  for {
     server <- startServerTask
     result <- originalTask.andFinally(stopServer(server))
  } yield result
}
```

Do you have a current example that compiles?  I'm curious to try for expressions for tasks but haven't been able to get this example to compile.

```
test in IntegrationTest := {
  executeInOrder(startServerTask, (test in IntegrationTest).taskValue, stopServerTask)
}
```

I think it's nice that this is possible for quick and dirty things, even though it doesn't clean up by running stopServer.  And I won't recommend it as a model for this case for that reason.  In the remote setup/deploy/run case it might be useful: historically we've had a bunch of tasks that don't return errors and are run in different combinations so sequence/executeInOrder might work well for those.

**Comment by jsuereth** [comment]
On Tue, Nov 26, 2013 at 2:23 PM, mighdoll notifications@github.com wrote:

> If the tests result is an error, then (test in IntegrationTest).value
> won't continue, because the result is an error, so my service is never
> stopped.
> 
> Error handling stops the whole task, and so later tasks don't run. OK, I
> get that and I see why that means the server might not be stopped if
> there's an error.
> 
> Would andThen/andFinally for tasks be the right fix for this? (for a
> workaround I realize we need use what we have)

andThen/andFinally already exists, in a limited form.

>  The ideal workaround is to have one piece of the task-puzzle encapsulate
> the logic:
> What you want to do is take this Task[_] which executes all tests, and tie
> in your 'start/stop' server into it.
> I wonder how much of a corner we're painted in when Task behavior is
> fragmented across keys though.
> 
> Tasks are the 'atoms' that sbt build customizers are effectively
> encouraged to use for everything (because they have access to settings). I
> suppose the core grammar of tasks is: run some in parallel, run some
> sequentially, attach task output to input, try/catch, attach task to key.
> I'd like to imagine we get to a complete grammar around tasks. Imagining
> I'm going to be explaining the task model, does that cover the core? Also,
> the current andFinally seems (so far to me) to require dropping to a
> different level of abstraction since it works on function not task.

Yeah, my hazy idea here is actually to create "pseudo" tasks for andFinally
behavior.... I.e. what if I could create a set of "bottom tasks" that are
added to the 'task request' if you encounter a certain task.

e.g. if I _ever_ see "startIntegrationServer" task show up, he can actually
append psuedo-tasks which get executed when the rest of execution is
completed.  Then he does his own cleanup, and only needs to be invoked
_before_ the test task, and automatically cleaned up at the end of the task
request.

The downside here is that you may need him cleaned up _BEFORE_ the end of a
task request, but this is at least a start.

> Do you have a current example that compiles? I'm curious to try for
> expressions for tasks but haven't been able to get this example to compile.
> 
> Ah right.  "taskValue" doesn't exist yet in sbt 0.13.0.  Here's the example
> form sbt in Action:

executeTests in IntegrationTest <<=

 (executeTests in IntegrationTest, uberJarRunner) apply {

   (testTask, runnerTask) =>

```
  for {

    runner <- runnerTask

    _ = runner.start()

    result <- testTask.andFinally(runner.stop())

  } yield result
```

 }

**Comment by mighdoll** [comment]
> "taskValue" doesn't exist yet in sbt 0.13.0.  Here's the example..

Ah, thanks.  I'll look forward to taskValue too, I see now that mark pushed an implementation a day or two ago.

**Comment by eed3si9n** [comment]
@harrah wrote:

> I'm not personally reimplementing `async` and then maintaining another large macro system either. These things are huge time sinks. So, I don't see it ever being implemented even if it were a somewhat reasonable idea.

One might read this and just go home, but I read this to be a "go ahead" for a plugin :) I've implemented the macro as sbt-sequential and documented it in [sequencing tasks with sbt-sequential](http://eed3si9n.com/sequencing-tasks-with-sbt-sequential).

**Comment by retronym** [comment]
> (Type hovers are wrong, navigate to declaration doesn't work. Known issues?)

http://assembla.com/spaces/scala-ide/tickets/1001449

We're going to fix this in Scala 2.11.

**Comment by retronym** [comment]
> A macro in the scala/async style has limitations as well. @jsuereth's example doesn't currently work in async because await within try/catch/finally isn't supported. (I don't know if it is fundamental- it might just not be implemented yet, but this style of macro is not at all trivial.) 

It's not fundamental, but non-trivial to get right. We should have it ready in January. (First attempt: https://github.com/scala/async/pull/11)

**Comment by harrah** [comment]
@eed3si9n Down the rabbit hole you go!

For expressions: `for(i <- 1 to n) yield println(someTask.value)` (currently crashes)

If expressions: `if(n > 1) someTask.value else 1` (currently always evaluates someTask)

Nested blocks: `{ { a.value; b.value }; { c.value; d.value} }`  (doesn't order within the nested blocks)

Access to the `var` needs to be synchronized or declared volatile.

If you haven't already taken a look at scala/async, you should.  It is very likely the same approach would be necessary here.

**Comment by eed3si9n** [comment]
As long as it crashes, I'm ok with it. I did think about non-shortcutting `if`-expression, and that's something I might take a look later esp taskDyn is designed to support that use case. Since it's a plugin I get to be more relaxed about the completeness.

**Comment by harrah** [comment]
I assume you mean you are ok with it crashing while in development or something?  My point is more that the reason for this plugin is to allow imperative code.  But, it really only allows a very restricted subset of Scala code.  sbt's macros make one change- inputs evaluated in parallel before the code executes.  This adds many restrictions, however.

**Comment by kell18** [comment]
It's just impossible to figure out exact way of doing it. Literally got 3 different ways to do it and NONE worked. Use case: override inputKey by `checking if the branch is correct then call inputKey otherwise fail`

**Comment by eed3si9n** [comment]
There's a how to guide here - https://www.scala-sbt.org/1.x/docs/Howto-Sequencing.html. If that doesn't work, please consult https://www.scala-sbt.org/community.html#how-can-I-get-help
